crypto/src/main/kotlin/web5/sdk/crypto/KeyGenerator.kt

```diff
@@ -0,0 +1,102 @@
+ * The `KeyGenerator` interface provides a blueprint for implementing cryptographic key
+ * generation and conversion functionalities for various cryptographic algorithms and key types.
+ *
+ * Cryptographic keys play a pivotal role in securing communication and data. This interface
+ * standardizes the key management operations including key generation, format conversion,
+ * and key derivation, ensuring that implementers adhere to a consistent API. In the realm
+ * of cryptographic applications, `KeyGenerator` enables developers to instantiate key pairs,
+ * convert keys between various formats, and derive public keys from private ones, while
+ * abstracting the specifics of the cryptographic algorithm and key type in use.
+ *
+ * Implementers are expected to provide concrete implementations for various key management
+ * activities for specified algorithms (e.g., RSA, EC) and key types (symmetric/asymmetric),
+ * and to handle potential exceptions, especially in scenarios such as deriving public keys
+ * in symmetric key contexts.
+ *
+ * ### Example Usage:
+ *
+ * ```
+ * val keyGenerator: KeyGenerator = ...
+ * val privateKey: JWK = keyGenerator.generatePrivateKey()
+ * val publicKey: JWK = keyGenerator.getPublicKey(privateKey)
+ * val privateKeyBytes: ByteArray = keyGenerator.privateKeyToBytes(privateKey)
+ * val restoredPrivateKey: JWK = keyGenerator.bytesToPrivateKey(privateKeyBytes)
+ * ```
+ *
+ * ### Note:
+ * For symmetric key generators, certain methods like `getPublicKey` or `publicKeyToBytes`
+ * may not be applicable and should throw an `UnsupportedOperationException`.
+ */
+public interface KeyGenerator {
+  /**  Indicates the algorithm intended to be used with the key. */
+  public val algorithm: Algorithm
+
+  /** Indicates the cryptographic algorithm family used with the key. */
+  public val keyType: KeyType
+
+  /** Generates a private key. */
+  public fun generatePrivateKey(options: KeyGenOptions? = null): JWK
+
+  /**
+   * Derives a public key from the private key provided. Applicable for asymmetric Key Generators only.
+   * Implementers of symmetric key generators should throw an UnsupportedOperation Exception
+   */
+  public fun getPublicKey(privateKey: JWK): JWK
+
+  /**
+   * Converts a private key to bytes.
+   */
+  public fun privateKeyToBytes(privateKey: JWK): ByteArray
+
+  /**
+   * Converts a public key to bytes. Applicable for asymmetric Key Generators only.
+   * Implementers of symmetric key generators should throw an UnsupportedOperation Exception
+   */
+  public fun publicKeyToBytes(publicKey: JWK): ByteArray
+
+  /**
+   * Converts a private key as bytes into a JWK.
+   */
+  public fun bytesToPrivateKey(privateKeyBytes: ByteArray): JWK
+
+  /**
+   * Converts a public key as bytes into a JWK. Applicable for asymmetric Key Generators only.
+   * Implementers of symmetric key generators should throw an UnsupportedOperation Exception
+   */
+  public fun bytesToPublicKey(publicKeyBytes: ByteArray): JWK
```

Comment:
I think we should only allow/expose dealing with JWKs. This makes things simpler to understand for developers, as they only need to understand a single concept.

That said, I'm not sure how these function will be used by clients, do you have an example?

If it is possible to reduce the API surface by removing them, I think that would be a win.